src/test/java/uk/gov/cshr/controller/emailUpdate/AgencyTokenVerificationControllerTest.java

```diff
@@ -0,0 +1,197 @@
+                .andExpect(model().attributeExists("emailUpdatedRecentlyEnterTokenForm"))
+                .andExpect(view().name("enterTokenSinceEmailUpdate"));
+
+        verify(identityService, times(1)).getDomainFromEmailAddress(EMAIL);
+    }
+
+    @Test
+    public void givenARequestToDisplayEnterTokenPageAndFormAlreadyExistsInModel_whenEnterToken_thenShouldDisplayEnterTokenPageWithAllPossibleOrganisationsAndTheExistingForm() throws Exception {
+        EmailUpdatedRecentlyEnterTokenForm existingForm = new EmailUpdatedRecentlyEnterTokenForm();
+        existingForm.setUid(IDENTITY_UID);
+        existingForm.setDomain(DOMAIN);
+        existingForm.setOrganisation(ORGANISATION);
+        existingForm.setToken(TOKEN);
+
+        mockMvc.perform(
+                get(VERIFY_TOKEN_URL + CODE)
+                        .with(CsrfRequestPostProcessor.csrf())
+                        .flashAttr("emailUpdatedRecentlyEnterTokenForm", existingForm)
+                        .flashAttr("uid", IDENTITY_UID)
+                        .flashAttr("domain", DOMAIN))
+                .andExpect(status().isOk())
+                .andExpect(model().size(5))
+                .andExpect(model().attribute("organisations", organisations))
+                .andExpect(model().attribute("domain", DOMAIN))
+                .andExpect(model().attribute("uid", IDENTITY_UID))
+                .andExpect(model().attribute("code", CODE))
+                .andExpect(model().attributeExists("emailUpdatedRecentlyEnterTokenForm"))
+                .andExpect(view().name("enterTokenSinceEmailUpdate"));
+    }
+
+    @Test
+    public void givenAValidTokenForm_whenCheckToken_thenShouldSubmitToken() throws Exception {
+        AgencyToken agencyToken = new AgencyToken();
+        agencyToken.setUid(AGENCY_TOKEN_UID);
+
+        when(csrsService.getAgencyTokenForDomainTokenOrganisation(DOMAIN, TOKEN, ORGANISATION)).thenReturn(Optional.of(agencyToken));
+        when(agencyTokenCapacityService.hasSpaceAvailable(agencyToken)).thenReturn(true);
+
+        EmailUpdate emailUpdate = new EmailUpdate();
+        emailUpdate.setEmail(EMAIL);
+        Identity identity = new Identity();
+        identity.setEmail(EMAIL);
+        when(emailUpdateService.getEmailUpdate(any(Identity.class), eq(CODE))).thenReturn(emailUpdate);
+        doNothing().when(emailUpdateService).updateEmailAddress(any(HttpServletRequest.class), eq(identity), eq(emailUpdate), eq(agencyToken));
+
+        mockMvc.perform(
+                post(VERIFY_TOKEN_URL + CODE)
+                        .with(CsrfRequestPostProcessor.csrf())
+                        .param("organisation", ORGANISATION)
+                        .param("token", TOKEN)
+                        .param("domain", DOMAIN)
+                        .param("uid", IDENTITY_UID)
+        )
+                .andExpect(status().is3xxRedirection())
+                .andExpect(redirectedUrl(REDIRECT_EMAIL_UPDATED));
+    }
+
```

Comment:
Add a unit test for the technical exception scenario.  (check redirected to log in page)  Probably should have an error message or a general error page.